ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,309 @@
+    assertNotFound();
+  });
+
+  test('Should reload when href template changes', async () => {
+    documentLayout = await buildLayout();
+    expect(documentLayout.element).to.exist;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // check that the layout updates when the document changes
+    documentLayout.hrefTemplate = '../dummy-layout.html';
+    const event = await waitForEvent(documentLayout, 'document-layout-changed');
+    expect(event.detail).to.exist;
+    const { element } = event.detail;
+    expect(element).to.exist;
+    expect(element.tagName).to.equal('DUMMY-LAYOUT');
+    // assert layout integrity
+    const { children } = element.shadowRoot;
+    expect(children).to.have.lengthOf(2);
+    expect(children[0].tagName).to.equal('SPAN');
+    expect(children[1].tagName).to.equal('NUXEO-INPUT');
+  });
+
+  test('Should reload when href base changes', async () => {
+    documentLayout = await buildLayout();
+    expect(documentLayout.element).to.exist;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // check that the layout updates when the document changes
+    documentLayout.hrefBase = `${base}/layout/`;
+    const event = await waitForEvent(documentLayout, 'document-layout-changed');
+    expect(event.detail).to.exist;
+    expect(event.detail.element).to.be.undefined;
+    // check 404 for unexisting layout
+    assertNotFound();
+  });
+
+  test('Should do input validation when the "validate" method is called', async () => {
+    documentLayout = await buildLayout();
+    // check that validation fails if required fields are not filled
+    expect(documentLayout.validate()).to.be.false;
+    const myTitle = 'My Title';
+    getWidget('Title').value = myTitle;
+    // assert that validation passes if required fields are filled
+    expect(documentLayout.document.properties['dc:title']).to.equal(myTitle);
+    expect(documentLayout.validate()).to.be.true;
+  });
+
+  test('Should do custom input validation when a "validate" method is available on the layout', async () => {
+    documentLayout = await buildLayout(
+      buildDoc({
+        'dc:title': 'My Title',
+        'dc:description': 'My Title',
+      }),
+    );
+    // check that validation fails if title and description are the same (custom validation)
+    expect(typeof documentLayout.element.validate).to.equal('function');
+    sinon.spy(documentLayout.element, 'validate');
+    expect(documentLayout.validate()).to.be.false;
+    expect(documentLayout.element.validate.calledOnce).to.be.true;
+    const myDescription = 'My Description';
+    // assert that validation passes if the values differ
+    getWidget('Description').value = myDescription;
+    expect(documentLayout.document.properties['dc:description']).to.equal(myDescription);
+    expect(documentLayout.validate()).to.be.true;
+    expect(documentLayout.element.validate.calledTwice).to.be.true;
```

Comment:
This is testing the layout implementation itself, which is ok but not sure if needed 🤷‍♂

Reply:
This is not testing the validate method of the `nuxeo-layout` but that of layout (html file) which is an input of this element.

I guess this is in a blurry area. If we accept that it is up to the `nuxeo-layout` to make sure it calls the `validate` method of the actual layout (if present), I guess we need to change this line (and the one in line 213) to at least:

```
expect(documentLayout.$.layout.validate.calledTwice).to.be.true;
```
Because the only thing our validate method does is calling the layout's `validate` method.

WDYT?